src/ring_theory/graded_algebra/radical.lean

```diff
@@ -0,0 +1,247 @@
+    decidable_pred (λ (i : ι), graded_algebra.proj 𝒜 i x ∉ I) := λ x, classical.dec_pred _,
+  letI : Π i (x : 𝒜 i), decidable (x ≠ 0) := λ i x, classical.dec _,
+  set set₁ := (graded_algebra.support 𝒜 x).filter (λ i, graded_algebra.proj 𝒜 i x ∉ I) with set₁_eq,
+  set set₂ := (graded_algebra.support 𝒜 y).filter (λ i, graded_algebra.proj 𝒜 i y ∉ I) with set₂_eq,
+  have set₁_nonempty : set₁.nonempty,
+  { rw finset.filter_nonempty_iff,
+    contrapose! rid₁,
+    rw ← graded_algebra.sum_support_decompose 𝒜 x,
+    apply ideal.sum_mem,
+    convert rid₁,},
+  have set₂_nonempty : set₂.nonempty,
+  { rw finset.filter_nonempty_iff,
+    contrapose! rid₂,
+    rw ← graded_algebra.sum_support_decompose 𝒜 y,
+    apply ideal.sum_mem,
+    convert rid₂, },
+  set max₁ := set₁.max' set₁_nonempty with max₁_eq,
+  set max₂ := set₂.max' set₂_nonempty with max₂_eq,
+  have mem_max₁ := finset.max'_mem set₁ set₁_nonempty,
+  have mem_max₂ := finset.max'_mem set₂ set₂_nonempty,
+  replace hxy : ∀ (i : ι), (graded_algebra.decompose 𝒜 (x * y) i : A) ∈ I := λ i, hI i hxy,
+  specialize hxy (max₁ + max₂),
+  have eq :=
+    calc  graded_algebra.proj 𝒜 (max₁ + max₂) (x * y)
+        = ∑ ij in ((graded_algebra.support 𝒜 x).product (graded_algebra.support 𝒜 y)).filter
+            (λ (z : ι × ι), z.1 + z.2 = max₁ + max₂),
+            (graded_algebra.proj 𝒜 ij.1 x) * (graded_algebra.proj 𝒜 ij.2 y)
+        : begin
+          rw [graded_algebra.proj_apply, alg_equiv.map_mul, graded_algebra.support,
+            graded_algebra.support, direct_sum.coe_mul_apply_submodule],
+          refl,
+        end
+    ... = ∑ ij in ((graded_algebra.support 𝒜 x).product (graded_algebra.support 𝒜 y)).filter
+            (λ (z : ι × ι), z.1 + z.2 = max₁ + max₂) \ {(max₁, max₂)} ∪ {(max₁, max₂)},
+            (graded_algebra.proj 𝒜 ij.1 x) * (graded_algebra.proj 𝒜 ij.2 y)
+        : begin
+            congr,
+            rw finset.sdiff_union_of_subset,
+            rintros z H,
+            rw finset.mem_singleton at H,
+            simp only [finset.mem_filter, H, finset.mem_product, eq_self_iff_true, and_true],
+            exact ⟨(finset.filter_subset _ _) mem_max₁, (finset.filter_subset _ _) mem_max₂⟩,
+        end
+    ... = ∑ (ij : ι × ι) in ((graded_algebra.support 𝒜 x).product
+            (graded_algebra.support 𝒜 y)).filter
+            (λ (z : ι × ι), prod.fst z + z.snd = max₁ + max₂) \ {(max₁, max₂)},
+            (graded_algebra.proj 𝒜 (prod.fst ij) x) * (graded_algebra.proj 𝒜 ij.snd y)
+        + ∑ ij in {(max₁, max₂)}, (graded_algebra.proj 𝒜 (prod.fst ij) x)
+            * (graded_algebra.proj 𝒜 ij.snd y)
+        : by { rw [finset.sum_union], apply finset.sdiff_disjoint }
+    ... = ∑ ij in ((graded_algebra.support 𝒜 x).product (graded_algebra.support 𝒜 y)).filter
+            (λ (z : ι × ι), z.1 + z.2 = max₁ + max₂) \ {(max₁, max₂)},
+            (graded_algebra.proj 𝒜 ij.1 x) * (graded_algebra.proj 𝒜 ij.2 y)
+        + _ : by rw finset.sum_singleton,
```

Comment:
I think [finset.sum_erase_add](https://leanprover-community.github.io/mathlib_docs/algebra/big_operators/basic.html#finset.sum_erase_add) will make this whole section a lot easier, as there's then no `singleton` set to mess around with.

Reply:
You'll of course need to replace `\ {(max₁, max₂)}` with `.erase (max₁, max₂)` for the rest of the proof.

Reply:
Thank you for the advice, I have changed everything to `erase`